File: docs/Cours/code/arbres.ml
type arbre = Nil | Noeud of int * arbre * arbre

let sauvegarde a f =
  let rec sauve a file =
    match a with
    | Nil -> output_string file "-1 "
    | Noeud (v, g, d) -> output_string file (string_of_int v ^ " ");
                         sauve g file;
                         sauve d file in
  let file = open_out f in
  sauve a file;
  close_out file


let chargement f =
  let rec charge lst file =
    let lst = if lst <> [] then lst else String.split_on_char ' ' (input_line file) in
    match lst with
    | [] -> failwith "probleme de format"
    | n :: lst' -> if n = "-1" then (Nil, lst)
                   else let g, lst2 = charge lst' file in
                        let d, lst3 = charge lst2 file
                        in (Noeud(int_of_string n, g, d), lst3) in
  let file = open_in f in
  let a = charge [] file in
  close_in file; a
          

let _ = let a = Noeud( 3, Noeud(5, Nil, Noeud(11, Nil, Nil)),
                                Noeud(43, Noeud(35, Nil, Nil), Nil) ) in
        sauvegarde a "toto"

File: docs/Cours/code/arbres.c
#include <stdio.h>
#include <stdlib.h>
#include "arbres.h"

void sauve(arbre a, FILE *file){
  if (a == NULL){
    fprintf(file, "-1 ");
  }  else {
    fprintf(file, "%d ", a->v);
    sauve(a->g, file);
    sauve(a->d, file);
  }
}

void sauvegarde(arbre a, char *f){
  FILE *file = fopen(f, "w");
  sauve(a, file);
  fclose(file);
}


arbre charge(FILE *file){
  arbre a;
  int n;
  fscanf(file, " %d", &n);
  if (n == -1){
    return NULL;
  }

  a = (arbre)malloc(sizeof(struct noeud));
  a->v = n;
  a->g = charge(file);
  a->d = charge(file);

  return a;
}

arbre chargement(char *f){
  arbre a;
  FILE *file = fopen(f, "r");
  a = charge(file);
  fclose(file);

  return a;
}

int main(void){
//  arbre a = chargement("toto");
  struct noeud a = {.v = 5, .g = NULL, .d = NULL},
	b = {.v = 13, .g = &a, .d = NULL},
	c = {.v = 17, .g = NULL, .d = NULL},
	d = {.v = 25, .g = &b, .d = &c};
  sauvegarde(&d, "titi");
  return 0;
}
